src/dmd/statementsem.d

```diff
@@ -3847,7 +3847,12 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
             }
             else if (t.ty == Tstruct)
             {
-                if (!ws.exp.isLvalue())
+                // https://issues.dlang.org/show_bug.cgi?id=22017
+                // if the with expression is an lvalue and a CallExp,
+                // the called function must be a ref function. In this
+                // situation we will still create the temporary so that
+                // we can destroy it after the WithStatement.
+                if (!ws.exp.isLvalue() || ws.exp.isCallExp())
```

Comment:
Shouldn't this rather be considered in `CallExp.isLvalue` s.t. it is consistent throughout dmd?

Reply:
I don't think that's going to work because ref functions are typically considered lvalues, so their address can be taken. If we modify this, a lot of code is going to break.

Reply:
It should obviously not apply to all calls to `ref` functions, just those that can be inferred to return a reference to a temporary.